### PR TITLE
Feature/285 schools lookup component

### LINKED ISSUE #285: Schools Lookup Component
As part of the new fundraising journey (https://github.com/comicrelief/technology-planning/issues/24), we need to build a component (molecule) to allow schools to lookup their addresses using their postcode.

This should use our school lookups api endpoint,

http://spa-lookups-apidocs.s3-website-eu-west-1.amazonaws.com/#api-General-EdCo_Lookup

and have much the same functionality as the current version in storybook

https://storybook.services.comicrelief.com/?knob-id=radioButtons4&knob-name=radiobuttons4&knob-label=Radio%20Buttons%3A%20not%20required%2C%20preselected&knob-Min%20Length=3&knob-Endpoint=https%3A%2F%2Flookups-staging.sls.comicrelief.com%2Fschools%2Flookup%3Fquery%3D&selectedKind=Schools%20Lookup&selectedStory=nothing%20is%20selected&full=0&addons=1&stories=1&panelRight=0&addonPanel=storybook%2Factions%2Factions-panel.

We should refactor the validation to use the Formik react library and use components from component library to build the component.

### PR DESCRIPTION
**Pull request by tomisaacmarshall**
Type: patch

Partially fixes #285 

## Changes proposed in this PR

This PR partially completes task 285 to recreate the old Storybook schools lookup component in the component library. 

I wanted to break it down so the PR wasn't too big (and because I was having a little trouble with the manual form part), so this includes the school lookup input only - the manual address entry / edit form will come later in a separate PR.

### New components

It adds three components:

- TextInputWithDropdown (atom) - this is a dumb component that just takes care of the visual side of things

- Typeahead (molecule) - this is a generic typeahead component that uses the above, and could be used for various lookup / search input fields. It adds the core functionality around fetching data in a debounced manner, managing / updating state, etc

- SchoolLookup (molecule) - this is the specific schools lookup version of the above component 

### Outstanding issues

- I don't have a great deal of testing experience, particularly in JS. It would be good to be able to test the Typeahead and SchoolLookup components more thoroughly, mocking the interactions etc, but wasn't really sure how to go about this - any advice would be very welcome!

- Accessibility is another area that I need to improve on. I'm not sure if TextInputWithDropdown  passes muster in this regard - again, grateful for any tips

- Do I need to do any more on styling? The dropdown is based on the original component; I wasn't working from any designs.

### CODE REVIEW

File: src/components/Atoms/TextInputWithDropdown/TextInputWithDropdown.style.js [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+import { screen } from '../../../theme/shared/size';
+
+const Container = styled.div`
+  position: relative;
+`;
+
+const Dropdown = styled.div`
+  font-family: ${({ theme }) => theme.fontFamilies('Montserrat')};
+  z-index: 1;
```
Comment (gusliedke) [inline src/components/Atoms/TextInputWithDropdown/TextInputWithDropdown.style.js:11]: Add `${zIndex('high')};` from _utils_

File: src/components/Atoms/TextInputWithDropdown/TextInputWithDropdown.style.js [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+`;
+
+const DropdownList = styled.ul`
+  list-style: none;
+  padding: 0;
+  margin: 0;
+`;
+
+const DropdownItem = styled.li`
+  line-height: 1.5rem;
```
Comment (gusliedke) [inline src/components/Atoms/TextInputWithDropdown/TextInputWithDropdown.style.js:32]: you can use our _spacing_ `${spacing('md)'}`

**Comment by AndyEPhipps** [comment]
> I don't have a great deal of testing experience, particularly in JS. It would be good to be able to test the Typeahead and SchoolLookup components more thoroughly, mocking the interactions etc, but wasn't really sure how to go about this - any advice would be very welcome!

It looks like we don't Cypress set up in CL yet, probably worth doing, right @gusliedke @KrupaPammi ?

File: src/components/Molecules/SchoolLookup/SchoolLookup.test.js [hunk 3]
```diff
@@ -0,0 +1,13 @@
+import React from 'react';
+import 'jest-styled-components';
+
+import renderWithTheme from '../../../hoc/shallowWithTheme';
+import SchoolLookup from './SchoolLookup';
+
+// todo: can we test the interactions?
```
Comment (gusliedke) [inline src/components/Molecules/SchoolLookup/SchoolLookup.test.js:7]: Usually we test interactions in our app. We can add some cypress tests in the integration part.

**Comment by KrupaPammi** [comment]
> > I don't have a great deal of testing experience, particularly in JS. It would be good to be able to test the Typeahead and SchoolLookup components more thoroughly, mocking the interactions etc, but wasn't really sure how to go about this - any advice would be very welcome!
> 
> It looks like we don't Cypress set up in CL yet, probably worth doing, right @gusliedke @KrupaPammi ?

I'm working on adding cypress tests on CL @AndyEPhipps

**Comment by gusliedke** [comment]
> > I don't have a great deal of testing experience, particularly in JS. It would be good to be able to test the Typeahead and SchoolLookup components more thoroughly, mocking the interactions etc, but wasn't really sure how to go about this - any advice would be very welcome!
> 
> It looks like we don't Cypress set up in CL yet, probably worth doing, right @gusliedke @KrupaPammi ?
> 
> > Accessibility is another area that I need to improve on. I'm not sure if TextInputWithDropdown passes muster in this regard - again, grateful for any tips
> 
> Probably worth installing the Lighthouse browser extension as a starting point

I think we could leave integration test on app level. What do u think?

**Comment by AndyEPhipps** [comment]
> > > I don't have a great deal of testing experience, particularly in JS. It would be good to be able to test the Typeahead and SchoolLookup components more thoroughly, mocking the interactions etc, but wasn't really sure how to go about this - any advice would be very welcome!
> > 
> > 
> > It looks like we don't Cypress set up in CL yet, probably worth doing, right @gusliedke @KrupaPammi ?
> > > Accessibility is another area that I need to improve on. I'm not sure if TextInputWithDropdown passes muster in this regard - again, grateful for any tips
> > 
> > 
> > Probably worth installing the Lighthouse browser extension as a starting point
> 
> I think we could leave integration test on app level. What do u think?

I think it's worth it; we've had a few problems end up on live that could've been avoided, although I guess specifically more related to snapshots etc. rather than purely functionality. 

Sounds like @KrupaPammi's on the case anyways 👍

**Comment by adamclark-dev** [comment]
@gusliedke makes sense to add tests at the component lib level. Now that component lib is going to be used in multiple applications, we don't want to have to be duplicating tests in multiple downstream applications.

**Comment by tomisaacmarshall** [comment]
Thanks for comments all. Have made your requested changes @gusliedke 

Have also just updated snapshots and added a commit message with the 'feat:' prefix.

@KrupaPammi do we need to wait for your integration tests before merging this?

**Comment by KrupaPammi** [comment]
> Thanks for comments all. Have made your requested changes @gusliedke
> 
> Have also just updated snapshots and added a commit message with the 'feat:' prefix.
> 
> @KrupaPammi do we need to wait for your integration tests before merging this?

@tomisaacmarshall please go ahead, I'm doing it on separate branch.

**Comment by tomisaacmarshall** [comment]
> Looks really decent @tomisaacmarshall, nice one. Only addition it needs, comparing to the existing Storybook component, is to actual render out the chosen details underneath.. unless I'm missing something?

I was thinking the lookup input component itself should just do the lookup + have some means to provide the selected school data to its parent - i.e. the onSelect prop. Then it's up to the parent to decide what to do with that data e.g. print it out.

Then I was planning to do a follow up PR that would add the address form as another molecule, plus an organism tying everything together, which will be the thing that includes all the functionality of the old component (including rendering the address of the chosen school). Does that sounds ok?

**Comment by AndyEPhipps** [comment]
> > Looks really decent @tomisaacmarshall, nice one. Only addition it needs, comparing to the existing Storybook component, is to actual render out the chosen details underneath.. unless I'm missing something?
> 
> I was thinking the lookup input component itself should just do the lookup + have some means to provide the selected school data to its parent - i.e. the onSelect prop. Then it's up to the parent to decide what to do with that data e.g. print it out.
> 
> Then I was planning to do a follow up PR that would add the manual address form (although it won't technically be a `<form>`...) as another molecule, plus an organism tying everything together, which will be the thing that includes all the functionality of the old component (including rendering the address of the chosen school). Does that sounds ok?

Okay, sweet; sounds like you've got it sussed, nice one 👍

**Comment by gusliedke** [comment]
> @gusliedke makes sense to add tests at the component lib level. Now that component lib is going to be used in multiple applications, we don't want to have to be duplicating tests in multiple downstream applications.

That's true! 👍

**Comment by AndyEPhipps** [comment]
@tomisaacmarshall : happy to let you add in those additional elements in separate issues and components if you feel like it's the right approach.

**Pull request merged**